cri-api policies (#566)
kubernetes/contributor-site: content/en/docs/code/cri-api-dev-policies.md

---
content_type: "reference"
title: Kubernetes feature development and container runtimes
weight: 55
---

> CRI is a plugin interface which enables kubelet to use a wide variety of container runtimes,

Reviewer: nit: do we have some doc guidelines whether we should use "enables kubelet" or "enables the kubelet"?

I prefer the former myself, but the latter seems to be used so often that I thought it was a convention that I just hadn't trained myself to follow...

Author: fixed here and in other document

without the need to recompile. CRI consists of a protocol buffers and gRPC API.
Read more about CRI API at [kubernetes docs](https://kubernetes.io/docs/concepts/architecture/cri/).

The mechanics of a feature development that requires new CRI APIs is covered in
documentation on CRI API [feature-development](/docs/code/cri-api-version-skew-policy#feature-development).

> This article declares policies for developing new Kubernetes features

Reviewer 2: nit
```suggestion
This article defines policies for developing new Kubernetes features
```

that require CRI API changes. The goal of these policies is to ensure great user
experience for people trying the new feature early, adopting it when it is
enabled by default, and relying on it as a GA functionality.

{{% alert title="Notice" color="warning" %}}

> This document establishes development policies for a new features requiring

Reviewer 2: nit
```suggestion
This document establishes development policies for new features requiring
```

> adherence to the policy by two currently participating runtime. This is not

Reviewer 2: nit
```suggestion
adherence to the policy by two currently-participating runtime. This is not
```

> a commitment from Kubernetes community to support these features. Also

Reviewer 2: nit
```suggestion
a commitment from the Kubernetes community to support these features. Also
```

> a commitment from Kubernetes community to support those these features. Also

Reviewer: ```suggestion
a commitment from Kubernetes community to support these features. Also
```

feature to work on two container runtimes: [containerd](https://containerd.io/)
and [CRI-O](https://cri-o.io/). These are two runtimes that are tested as part
of a Kubernetes development and release process to refer features portability.
When this document refers to two container runtimes, it assumes both -
containerd and CRI-O. If any other container runtimes begin working actively
with the Kubernetes community, or one of existing container runtimes will stop
engagement, this document will be updated.

## Same maturity level (for beta and GA)

Implementation of an API needed for a Kubernetes feature in a container runtime
**must** be at least the same maturity as in k8s at a moment of Kubernetes release.
This is similar to the [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior)
when one feature is replacing another.

With containerd and CRI-O it means that for the GA features, there should be a
release of containerd and CRI-O implementing APIs needed for the feature in
those container runtimes. And those APIs **must not** be marked as experimental

> features ([containerd experimental features](https://containerd.io/releases/#experimental-features)).

Reviewer 3: This means that the containerd project can actively influence the development in Kubernetes. It should be the other way around from my point of view: Kubernetes declares a feature to GA and container runtimes will adapt.

Author: It is a process and we have decided that we are working with two runtimes today. Either can block the feature. So far it was based on a real concern, which makes things better for everybody. So this is very much intentional

> For the beta, neither container runtime has a notion of beta feature or release and

Reviewer 2: Speaking with my containerd hat on:

* We do not have betas for features but we do mark certain features as experimental (called out above)
* containerd does have beta releases and RCs for new minor versions, but the beta and RC periods are relatively short.  We are formalizing some policy changes around this in https://github.com/containerd/containerd/pull/11294 now

I don't think this needs to change the general policy for features in Kubernetes, but we may want to clarify that the notion of "beta" is not equivalent between Kubernetes and containerd.

Reviewer 4: In general, we could emphasize that external projects have their own ideas and jargon concerning stability (or not).

> realistically the same maturity criteria applies as for GA.

Reviewer 3: Beta graduations may come with additional CRI changes. I assume container runtimes should adapt to them and document accordingly which feature graduation state they support.

Author: yes. Do you feel this needs to be spelled out explicitly?

## Same maturity level (for alpha)

There should be at least one implementation of an API needed for the Kubernetes

> feature merged into the container runtime default branch (main) or marked as

Reviewer 2: possible nit?
```suggestion
feature merged into the container runtime development branch (main) or marked as
```

> experimental. An e2e test may demonstrate that the feature is working should be
> merged alongside the code. Note, tests may run against the default branch of a

Reviewer 2: "may" and "should" are contradicting each other here.

```suggestion
experimental. An e2e test which demonstrates that the feature is working should be
merged alongside the code. Note, tests may run against the default branch of a
```

API needed for a Kubernetes feature before this Kubernetes feature can be
promoted to GA.

## Safe Kubernetes defaults

The feature cannot be enabled by default in Kubernetes as a beta feature before
the required APIs are implemented in both container runtimes (containerd and
CRI-O) and there is a GA release of a container runtime for each. The feature
can be marked as beta, but disabled by default, if there is only one container
runtime implementation of a required API that is released as GA. Note, as for
any Kubernetes features, at least one release with the beta feature enabled by
default is required before it is progressing to GA.

## Guarantee portability

The feature design (KEP PR) MUST be lgtm-ed by container runtime maintainers of
CRI-O and containerd.

The feature can only be merged as alpha in Kubernetes, if there is an agreement
from both container runtime maintainers on the feature design in general and API
shape. Gaining this agreement will often involve authoring the pull request
demonstrating an API implementation to the both container runtime repositories
or an alternative way for container runtime maintainers to confirm viability of

> suggested CRI APIs.

Reviewer 3: Hm, would it be enough to state that the CRI API change has outlined in detail in the KEP? I'm not sure what the benefit is of bringing the API into the projects without the actual functionality.

Author: we need an agreement from runtime that this API is viable. Otherwise we may end up with the CRI API that is not implemented by any or one of runtimes. Getting review from two runtimes gives some confidence that it will be viable for other possible runtimes and will be portable enough

API) and **must** not depend on the cluster or node configuration that is not
clearly exposed via these APIs. For example, it is OK to have Windows specific
features that are configured via Pod API. But it is not OK to design a feature
that will work on one container runtime and incompatible with the other
container runtime. There are three exceptions to this case:

- there will be a different behavior during the feature adoption period while
  older runtime versions do not support the API yet. In those cases, attempting
  to try the feature **must** result in failing as fast as possible.  
- LTS and older versions of container runtimes may not have an implementation of
  an API and still be widely used by Kubernetes end users.  
- If any of container runtime underlying systems cannot support the feature
  in-principle (e.g. [kata containers](https://katacontainers.io/) with CRI-O
  may have limitations), while CRI-O still supports the feature without these
  systems configured, this **must** be designed as part of a normal operation.
  In this case, Pod or Node APIs **must** handle these cases gracefully and
  those **must** be documented clearly.

> 

Reviewer 3: Do we want to add general development criteria for the CRI? For example:

- Documenting intended behavior directly in the protocol definition to simplify runtime adoption.
- Simplify the intended change as much as possible.
- Reusing existing fields only if their logical meaning allows it and does not infer with existing features.

Author: yes, those likely belong to other document. Added a section "Designing new CRI APIs" to the version skew policy document
